File: backend/src/routes/media.ts

import express from 'express';
import multer from 'multer';
import path from 'path';
import fs from 'fs';
import { extractAudioFromMedia } from '../services/ffmpegService';
import { transcribeAudio } from '../services/asrService';

const router = express.Router();

// Configure multer for media uploads
const storage = multer.diskStorage({
  destination: (req, file, cb) => {
    const uploadDir = path.join(__dirname, '../../uploads/media');
    if (!fs.existsSync(uploadDir)) {
      fs.mkdirSync(uploadDir, { recursive: true });
    }
    cb(null, uploadDir);
  },
  filename: (req, file, cb) => {
    const uniqueSuffix = Date.now() + '-' + Math.round(Math.random() * 1E9);
    cb(null, file.fieldname + '-' + uniqueSuffix + path.extname(file.originalname));
  }
});

const upload = multer({ storage });

// Upload media file
router.post('/upload', upload.single('file'), async (req, res) => {
  try {
    if (!req.file) {
      return res.status(400).json({ error: 'No file uploaded' });
    }
    // Step 1: Extract audio from uploaded media
    const uploadDir = path.join(__dirname, '../../uploads/media');
    const mediaPath = req.file.path;
    let audioPath = null;
    let transcript = null;
    try {
      audioPath = await extractAudioFromMedia(mediaPath, uploadDir);
      // Step 2: Transcribe audio using ASR
      transcript = await transcribeAudio(audioPath);
    } catch (processingError) {
      console.error('Media processing error:', processingError);
    }
    res.json({
      mediaId: Date.now().toString(),
      status: 'processed',
      fileName: req.file.originalname,
      fileSize: req.file.size,
      fileType: req.file.mimetype,
      audioPath: audioPath ? path.basename(audioPath) : null,
      transcript: transcript || null
    });
  } catch (error) {
    console.error('Media upload error:', error);
    res.status(500).json({ error: 'Media upload failed' });
  }
});

export default router;

File: services/ai/app/api/v1/media.py

from fastapi import APIRouter, UploadFile, File, HTTPException
from typing import Optional
import shutil
import os
from ...services.ffmpeg_service import extract_audio_from_media
from ...services.asr_service import transcribe_audio

router = APIRouter()

UPLOAD_DIR = os.path.join(os.path.dirname(__file__), '../../../uploads/media')
os.makedirs(UPLOAD_DIR, exist_ok=True)

@router.post('/upload')
def upload_media(file: UploadFile = File(...)):
    try:
        file_path = os.path.join(UPLOAD_DIR, file.filename)
        with open(file_path, 'wb') as buffer:
            shutil.copyfileobj(file.file, buffer)
        audio_path = None
        transcript = None
        try:
            audio_path = extract_audio_from_media(file_path, UPLOAD_DIR)
            transcript = transcribe_audio(audio_path)
        except Exception as processing_error:
            print(f'Media processing error: {processing_error}')
        return {
            'mediaId': str(int(os.path.getmtime(file_path))),
            'status': 'processed',
            'fileName': file.filename,
            'fileSize': os.path.getsize(file_path),
            'fileType': file.content_type,
            'audioPath': os.path.basename(audio_path) if audio_path else None,
            'transcript': transcript or None
        }
    except Exception as e:
        raise HTTPException(status_code=500, detail=f'Media upload failed: {e}')
